CMakeLists.txt

```diff
@@ -1114,12 +1100,12 @@ add_library_variant(
     STACK_SIZE 4K
 )
 add_library_variant(
```

Comment:
Looks like we've reordered
```
v7-m soft nofp
v7-m softfp fpv4-sp-d16
v7-em soft nofp
```
to
```
v7-m soft fpv4_sp_d16
v7-m soft nofp
v7-em soft nofp
```

This is explained by
```
When no -mfpu=none is specified, the compiler internally adds all other
possible fpu settings when searching for matching variants. So for the
no-fpu variant to win, it has to be added to multilab.yaml after all
other fpu variants. So reorder the add_library_variant calls accordingly.
```
I think it will be worth adding a comment to the CMake file, possibly above the first add_library_variant or here.

I'm also trying to work out if there is a problem with the selection code in the compiler. If there is, we would want to mark the comment with a FIXME: This is a workaround that can be removed when the selection logic is fixed.

The following is what I'd expect from the compiler, I don't think that we'll be able to fix in this patch.

Thinking about the selection logic there are 2 parts, not helped by the values of the mfloat-abi=softfp option which conflates the two:
* Can I support hardware floating point instructions? (controlled by -mfpu or inferred from -mcpu/-march)
* Can I pass floating point arguments in floating-point/SIMD extended registers? (ostensibly controlled by -mfloat-abi)

In an ideal world the two options would be independent of each other, but instead they are linked by a bunch of opaque implications.

I would expect that when there is no instance of -mfpu=none
* If the cpu/arch has no floating point unit or SIMD registers (integer only MVE) then the only possible answer is soft_nofp.
* If the cpu/arch has no floating point, but does have SIMD registers (integer only MVE) then there are two possible answers: soft_nofp and hard_nofp, with the latter selected by -mfloat-abi=hard.
* If the cpu/arch has floating point then there are 3 possible answers: soft_nofp with -mfloat-abi=soft, softfp with -mfloat-abi=softfp, and hard with -mfloat-abi=hard.

I would expect that with an instance of -mfpu=X then the compiler driver should add all variants that are "less than, i.e. compatible with X floating point units". The exception being:
* if -mfloat-abi=soft then do not pass through any floating point units (i.e. same as -mfpu=none)

I would expect that with no instance of -mfpu then the compiler driver should look up the default FPU from the CPU/Arch, and treat it in the same way as above.

We may need to look at what value of -mfloat-abi is being passed through. If there is no FPU then:
* softfp should never be passed, either an error or convert it to soft.
* hard is permitted if the CPU has SIMD registers (integer only MVE) otherwise error or conver to soft.

Reply:
Actually we reordered
```
armv7m soft nofp
armv7m soft fpv4_sp_d16
armv7em soft nofp
armv7em hard fpv4_sp_d16
armv7em hard fpv5_d16
```
to
```
armv7m soft fpv4_sp_d16
armv7m soft nofp
armv7em hard fpv4_sp_d16
armv7em hard fpv5_d16
armv7em soft nofp
```
I think the important thing is to have the nofp variant to be the last variant added for that particular CPU. So I'll add two comments before both of the `add_library_variant` calls that add the nofp variants.